### ISSUE TITLE
Content Negotiation & Accept headers with q-values

### ISSUE THREAD

**Issue by vanHoesel** [PerlDancer/Dancer2#712]
A proposal to make writing better API servers with Dancer2...

As I recently have worked with Dancer2 and trying to use it as a REST or RESTful API-server, I encountered some troubles with content-negociation. My experience is that doing it right with Dancer is taking too much effort to get choose what do for each possible accepted content variant. This goes way beyond the existing serialisers or for the sake of completeness, beyond `Dancer2::Serializer::Mutable`. Not only does it need fixing to handle the different q-values but it also is not easy to make a difference between 'collection+json', 'application/json' or 'hal+json' and neither will it handle things like different file formats as 'image/jepg' or 'application/pdf' as being served as dynamic content. And beyond that, it only deals with HTTP-Accept MIME-types, not with language or encoding nor charset.

So this is what I'd like to see:

```
choose Accept (
    text_html
        => { # deal with HTML },
    application_json => {
        my %data = do_some_stuff_with( param('id') );
        return to_json %data;
    },
    'application/xml' => {
        my %data = do_some_stuff_with( param('id') );
        return to_xml %data;
    },
    ['image/jpeg', 'image/png']
        => { # make nice images },
    { %options }
);
```

what it should do, is execute the different code-REFs depending on what is handed through the HTTP-Accept header (an no other fancy wizzadry) and deal with the q-values.

Naturally, the Response HTTP Content-Type can be set by this mechanism, since you also know where to choose from.

This might look as more work when there is something like Mutable, but that does not give fine grain control over the tons off MIME-types there are. And then there is of course the list of options below.

The options could include:
- `default => $MIME_type` which could be used if the Accept header provides a `*/*`
- `Not-Acceptable => {…}` to handle situations where there is no possible match, usually a `Status: 406` can be treated as a good default, but returning `Status: 300` and a list of links to alternates would be very nice, especially if that list can be generated from list of acceptable types.

This whole mechanism works for the other Accept headers as well and can be nested.
- `choose Accept`
  and
- `choose Accept_Language`
- `choose Accept_Charset`
- `choose Accept_Encoding`

Please... see this as a RFC, which I might implement myself with lot of your help and have some nice exercise in writing plugins, or you might be better off coding it with the team.

**Comment by rleir** [comment]
Agreed that this is best done in a plugin. Maybe a route decorator plugin. 

This issue is stale. Close it? When you get a moment, thanks!

**Comment by vanHoesel** [comment]
it is not 'stale' during the Londen Perl Workshop i will tel more about it. There is one nasty problem with the plugin I am developing, the underlying HTTP::Headers::ActionPack module by Steven Little and Dave Rolsky is not RFC compliant - it gets quite OKish but it made me wonder if I would use faulty modules to implement the plugin, send in a bug report - or just write my own ContentNegotiation plugin

**Comment by SysPete** [comment]
I think @rleir's view is **not** that your requirements and ideas are stale but that since there has been no update for > 12 months and agreement seemed to have been reached that a plugin is the correct route then this ticket should be closed.

@vanHoesel do you have a public repo for your new plugin? Perhaps you could add a link in a comment to this issue so these things stay connected and then close the issue?

**Comment by vanHoesel** [comment]
Okay - it needed to be done anyways - but after writing hours on POD, this is what the module more or less would look like: [HTTP::ContentNegotiation]( https://github.com/THEMA-MEDIA/Dancer2-Plugin-HTTP-ContentNegotiation). Hopefully the Synopsis is clear enough.

**Comment by rleir** [comment]
@vanHoesel Very nice. Will it be more than an example? I will use get_language most likely. Then, with HTTP 'Accept-Language' and the browser's locale setting, we can be multilingual (is this the best approach?)
@SysPete, yes, exactly.
Thanks -- Rick

**Comment by vanHoesel** [comment]
@rleir I can not say if it is the best approach. Although I wrote this with REST-api's in mind, HTTP::ContentNegotiation is not limited to api's. It is certainly 'a way' to serve a web-page or an api-response in a language defined by the browser or clients preferences. More difficult it might become when you want the end user to give the ability to select a language. You would had to manually forge the HTTP Request to include a different Accept-Language . . but certainly for first-time-visitors, the Home page could be served in the language of the browser.

**Comment by veryrusty** [comment]
@vanHoesel++ that looks nice! I'll make further suggestions over on your repo!

**Comment by cromedome** [comment]
> I think [@rleir](https://github.com/rleir)'s view is **not** that your requirements and ideas are stale but that since there has been no update for > 12 months and agreement seemed to have been reached that a plugin is the correct route then this ticket should be closed.

That was how I was taking it.

Closing this, but happy to reopen this again if necessary. Thanks everyone!

**Comment by vanHoesel** [comment]
no updates for > 10 years 😊
